text/0000-extension-layer.md

```diff
@@ -0,0 +1,90 @@
+# Meta
+[meta]: #meta
+- Name: Add extension layer to exchange data
+- Start Date: 2023-10-09
+- Author(s): [c0d1ngm0nk3y](https://github.com/c0d1ngm0nk3y), [pbusko](https://github.com/pbusko)
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request:
+- CNB Pull Request:
+- CNB Issue:
+- Related: [RFC#0105 Support Dockerfiles](https://github.com/buildpacks/rfcs/blob/main/text/0105-dockerfiles.md)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This RFC introduces support for Extension Layers to allow data transfer between the build environment and the Kaniko execution.
+
+# Motivation
+[motivation]: #motivation
+
+This change allows extensions to possess their own layers to utilize during the generation/extend process. Additionally, it ensures that extension output does not inadvertently interfere with other extension or buildpack layers during the build, and it does not unintentionally become part of the final application image.
+
+This would allow distroless run images to be extended.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This follows up on RFC-0105 and proposes that during the execution of the extension's `bin/generate`, an extension is allowed to write arbitrary data to its exclusive layer. This data then becomes accessible during the execution of the `extend` phase via Kaniko context. The content of these extension-specific layers is ignored at build and launch time, it serves only the extension phase.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Before execution of the `bin/generate`, the lifecycle will create a distinct writable layer for each extension which passed detection. The extensions can then write to these layers, and the Kaniko context is set to the corresponding layer during the `extend` phase instead of the `<app>` directory. The Extension Layers will not be included in the final image by the lifecycle.
+
+The location of the Extension Layer will be provided to the `bin/generate` via additional environment variable `CNB_EXT_LAYER_DIR`.
+
+# Migration
+[migration]: #migration
+
+Since we would change the Kaniko context, this would be breaking for existing extensions.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
```

Comment:
This might make build.Dockerfiles and run.Dockerfiles from extensions less portable - for example at the moment, it doesn't matter whether the build engine is `docker build` or `kaniko` but having Dockerfiles expect volumes to be present might change that.

I could be wrong, but there doesn't seem to be a transparent way to mount in volumes for `docker build` without adding Dockerfile instructions: https://docs.docker.com/build/guide/mounts/#add-a-cache-mount ...this would preclude us from doing something like [using the daemon for build image extension](https://github.com/buildpacks/pack/pull/1791) to improve performance.

Reply:
This RFC does not implies volumes usage, but rather build context:

- https://github.com/GoogleContainerTools/kaniko#kaniko-build-contexts
- https://docs.docker.com/build/building/context/

Reply:
Got it. In that case, if we want to make both the "extension layers" directory and the app directory available in the build context, these would need to be children of a parent directory that would be prepared by the platform, correct? Could you add some illustrative details to the RFC text?